IPs/RSKIP106.md

```diff
@@ -0,0 +1,135 @@
+# Precompiled contract with assorted BTO utility functions
+
+|RSKIP          |106 |
+| :------------ |:------------- |
+|**Title**      |Precompiled contract with assorted BTO utility functions |
+|**Created**    |01-FEB-19 |
+|**Author**     |AM |
+|**Purpose**    |Usa |
+|**Layer**      |Core |
+|**Complexity** |1 |
+|**Status**     |Draft |
+
+## Abstract
+
+A new precompiled contract (i.e., native, hardwired onto the RSK consensus layer) is introduced. This contract contains a few complex BTO-related operations that help leverage some of the work BTO contracts need to repeatedly perform.
+
+## Motivation
+
+With the arising need of BTOs (ICOs over Bitcoin) on top of the RSK network, a need for different BTO token and/or token-emitting contracts to perform certain complex BTO-related operations on a regular basis is in place. If written in solidity (or assembly WOLOG) and executed on the RVM, each of these functions would be expensive - and even more expensive to write and test every time. Furthermore, different implementations can lead to errors potentially protruding to loss of funds. A well-written, tested and precompiled function set that covers these operations will leverage the complexity on the BTO contracts' side while at the same time enforcing a certain level of security by implementing these as part of the RSK consensus.
```

Comment:
What do you mean with "test every time" ? It it's a library then it can be tested just once.

Reply:
I was thinking about different people writing their own BTO contracts and potentially using different libraries for this, or even writing their own. It is true, if you reused an existing solidity library for this, you wouldn't have to retest things. That said, I think that model doesn't actually enforce that reuse, but rather makes it an optional choice within the development process.